src/evaluators/UnaryExpression.js

```diff
@@ -68,6 +68,10 @@ function evaluateDeleteOperation(expr: Value | Reference, realm: Realm) {
 
     // b. Let baseObj be ! ToObject(GetBase(ref)).
     let base = Environment.GetBase(realm, ref);
+    if (base instanceof AbstractValue && !(base instanceof AbstractObjectValue)) {
+      // if the base is abstract but not an abstract object, we have a type error
```

Comment:
No! It might be an object at runtime.

Reply:
Please can you explain?

Reply:
If base is an AbstractValue with type Value (Top), we do not know that it will not be an object at runtime.

Reply:
In other words, `AbstractObjectValue` represents that we *know* it is an object whereas `AbstractValue` means we're not sure what it is (but it could as well be an object)?

Reply:
What if we didn't throw a TypeError, but another type of error? It's far better than creating an invariant, which we can't recover from.

Reply:
We can't evaluate this expression at compile time, so we either have to produce a compile time error rejecting this code, or the operation needs to result in an abstract value that will do the right thing (including throwing a type error) at runtime.

Reply:
Oh, and to be clear: You should never write an invariant unless you firmly believe that it can never be false. You should always document the reasons for your belief. Should the invariant fail, that should be because there is a bug, not because of some bad input.

Reply:
@hermanventer Would there be a good PP code that I could use that has already been made? I looked through the list but wasn't sure what you'd say is best.